pages/api/retrieval.ts

```diff
@@ -0,0 +1,112 @@
+    let submissionIDlist: (string | undefined)[] | undefined = [];
+    try {
+      //Create dynamodb client
+      const db = new DynamoDBClient({ region: process.env.AWS_REGION ?? "ca-central-1" });
+      //Create form responses db param
+      //NB: A filter expression is applied after a Query finishes, but before the results are returned.
+      //Therefore, a Query consumes the same amount of read capacity, regardless of whether a filter expression is present.
+      const getItemsDbParams = {
+        TableName: "Vault",
+        IndexName: "retrieved-index",
+        Limit: parseInt(maxRecords as string),
+        //Cannot use partitin key or sort key such as formID and Retrieved attribute in keyConditionExpression simultaneously
+        KeyConditionExpression: "Retrieved = :zero",
+        ExpressionAttributeValues: { ":formID": { S: formID }, ":zero": { N: "0" } },
+        //A filter expression cannot contain partition key or sort key attributes.
+        //You need to specify those attributes in the key condition expression, not the filter expression.
+        FilterExpression: "FormID = :formID",
+        ProjectionExpression: "FormID,SubmissionID,FormSubmission,Retrieved",
+      };
+      //Get form's responses for formID
+      const formResponses = await db.send(new QueryCommand(getItemsDbParams));
+      //Collecting items submissionIDs for logging and updating items.
+      submissionIDlist = formResponses?.Items?.map((response) => {
+        return response.SubmissionID.S;
+      });
+      if (formResponses && formResponses.Count && formResponses.Count > 0) {
+        //Preparing a query to set records's attribute Retrieved to 1.
+        for (const submissionID of submissionIDlist as string[]) {
+          //Create a new statement object
+          const updateItem = {
+            TableName: "Vault",
+            Key: {
+              SubmissionID: {
+                S: submissionID,
+              },
+              FormID: {
+                S: formID,
+              },
+            },
+            UpdateExpression: "SET Retrieved = :retrieved",
+            ConditionExpression: "SubmissionID = :submission",
+            ExpressionAttributeValues: {
+              ":retrieved": { N: "1" },
+              ":submission": { S: submissionID },
+            },
+            ReturnValues: "NONE",
+          };
+          //Update one item at the time
+          await db.send(new UpdateItemCommand(updateItem));
+        }
+      }
+      //Return form data
+      return res.status(200).json({ responses: formResponses.Items });
+    } catch (error) {
+      //logging submissionIDs
+      logMessage.warn(`Possible data loss for these submissionIDs`);
+      for (const subID in submissionIDlist) logMessage.warn(subID);
+      logMessage.error(error);
+      return res.status(500).json({ responses: "Error on Server Side" });
+    }
+  }
+}
+export default isRequestAllowed(["GET"], validate(retrieval));
```

Comment:
Unfortunately the `validate()` wrapper will not correctly evaluate the token on this route.  The `validate()` function checks against the bearer token however this route receives the temporary token which contains the `email

Reply:
A new function will need to be written that takes the temporary token and checks against the database to see if:
1. The token can be verified: `jwt.verify(()` and contains an email address
2. The token has not expired
2. The token exists in the Database
3. The token record retrieved from the database includes the email, active, and template_id fields
4. The email on the token record matches the email in the jwt.verify() payload.
5. The active flag boolean is true.